resolver/resolver.cc

```diff
@@ -3917,6 +3917,83 @@ class ResolveSanityCheckWalk {
     }
 };
 
+class CheckAbstractClassInitialization {
+public:
+    struct CheckAbstractClassInitializationWalkResult {
+        vector<ast::ParsedFile> trees;
+    };
+
+    ast::ExpressionPtr postTransformSend(core::Context ctx, ast::ExpressionPtr tree) {
+        auto &send = ast::cast_tree_nonnull<ast::Send>(tree);
+        if (send.fun != core::Names::new_()) {
+            return tree;
+        }
+
+        auto *id = ast::cast_tree<ast::ConstantLit>(send.recv);
+        if (id == nullptr || !id->symbol.exists() || !id->symbol.isClassOrModule()) {
+            return tree;
+        }
+
+        auto symbol = id->symbol.asClassOrModuleRef().data(ctx);
+        auto singletonClass = symbol->lookupSingletonClass(ctx.state);
+
+        if (!singletonClass.exists()) {
+            return tree;
+        }
+
+        if (symbol->flags.isAbstract) {
```

Comment:
Structuring the code like this would be slightly better:

```suggestion
        auto symbol = id->symbol.asClassOrModuleRef().data(ctx);
        if (!symbol->flags.isAbstract) {
            return tree;
        }
        auto singletonClass = symbol->lookupSingletonClass(ctx.state);
        if (!singletonClass.exists()) {
            return tree;
        }

        ....
```